### DISCUSSION TITLE
`process.cwd()` gives different results when called with different package managers.

### DISCUSSION THREAD

**Discussion by btea** [nodejs/help#4630] [Q&A]
I started a terminal locally, as shown below. (Note that the path starts with lowercase.)

![image](https://github.com/user-attachments/assets/7ff2c370-888b-442c-b0d0-581a989a8a98)

I wrote a simple code test as shown below.

![image](https://github.com/user-attachments/assets/8ed201b0-e7e9-439e-8cfa-fb21d5089ed2)

When I start the script using `pnpm` and `npm`, I get two different results. I noticed that when I started it with `pnpm`, the first line of information printed had an additional capitalized path compared to when I started it with `npm`.
![image](https://github.com/user-attachments/assets/3689cf42-9d29-4521-80dd-6d191bad1490)

I tried debugging the code, but I couldn't see the implementation inside `rawMethods.cwd()`.
```js
const rawMethods = internalBinding('process_methods');

function wrappedCwd() {
  if (cachedCwd === '')
    cachedCwd = rawMethods.cwd();
  return cachedCwd;
}
```

If you have time, please take a look, thank you.  @RedYetiDev

**Comment by avivkeller** [comment]
What happens when you just run the file? With no package managers?

**Reply by btea** [reply]
As shown in the figure below, the path starting with lowercase is printed directly.

![QQ_1723073904141](https://github.com/user-attachments/assets/76766157-7a9d-4499-b6ed-ddc546a0eb99)

**Reply by avivkeller** [reply]
So this may be a PNPM issue?

**Reply by btea** [reply]
I'm not sure, I'll try to check out pnpm, thanks for your reply.

**Comment by btea** [comment]
I encountered another problem about `process.argv`. I wrote a test file test.mjs with the following content:
```js
console.log(process.argv.slice(2));
```
When I execute `node test.mjs config set save-prefix '' --location=project` in different terminals, I get different results.
|cmd|powershell|
|---|---|
|[ 'config', 'set', 'save-prefix', "''", '--location=project' ]|[ 'config', 'set', 'save-prefix', '--location=project' ]|

Could you please help me take a look? Thanks. @avivkeller

**Comment by Felipeness** [comment]
Windows being Windows. NTFS is case-insensitive but case-preserving, so the actual stored path might be `C:/Users/...` but if a process does `chdir("c:/users/...")`, then `GetCurrentDirectoryW` returns the lowercase version — whatever casing was used when the directory was last set.

The difference you're seeing comes from how npm and pnpm spawn child processes:

- **npm** inherits the shell's current directory as-is (lowercase in your case)
- **pnpm** resolves or normalizes the path before spawning, which changes the casing to match the actual filesystem casing (`C:/Users/...`)

Neither is wrong. Both paths point to the same directory.

If you need consistent casing, `realpathSync.native` gives you the actual filesystem casing:

```js
import { realpathSync } from "node:fs";

const canonicalCwd = realpathSync.native(process.cwd());
```

Or for path comparison, just go case-insensitive on Windows:

```js
import path from "node:path";

function samePath(a, b) {
  return path.resolve(a).toLowerCase() === path.resolve(b).toLowerCase();
}
```

Regarding your follow-up about `process.argv` differences between CMD and PowerShell — also not a Node issue. CMD passes `''` as a literal two-character string, while PowerShell interprets `''` as an empty string and removes it from the argument list. Use double quotes in PowerShell to pass an empty string argument. Standard shell quoting behavior on Windows, ran into this before myself.

**Comment by Errr0rr404** [comment]
`process.cwd()` is the **process current working directory**, not “where the entry file lives”. Different package managers / how you launch the script can start Node with different CWDs.

Typical causes:
- Running via `npm test` / `pnpm` / `yarn` from repo root vs `node path/to/file.js` from another folder
- IDE run configs with a custom working directory
- Case-insensitive macOS paths displaying differently (`/Users/...` vs shell cwd)

What to use instead when you mean “directory of this package/file”:

```js
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
// stable regardless of cwd
```

Or for the package root, walk up to the nearest `package.json` from `__dirname`.

So: normalize on `__dirname` / `import.meta.url` for path resolution; use `process.cwd()` only when you truly want the user’s launch directory.

**Comment by k4hvecii** [comment]
The output from process.cwd() actually seems to be the same in both cases: c:\a.

The extra C:\a shown when running pnpm test is not produced by process.cwd(). It is part of pnpm's lifecycle script output.

With npm you get something like:

npm:
a@1.0.0 test
node z.mjs
c:\a

While pnpm prints the package directory in its script header:

pnpm:
a@1.0.0 test C:\a
node z.mjs
c:\a

So the Node.js process itself is still reporting c:\a in both cases.

The difference between C:\a and c:\a is also not significant on Windows because Windows paths are normally case-insensitive.

Therefore I don't think this is caused by the implementation of process.cwd() or rawMethods.cwd(). It looks like a formatting difference between npm and pnpm's lifecycle output.

If you want to verify it independently of the package manager output, you can run:

node -e "console.log(process.cwd())"

from the same directory and compare that result.
